[#41816] Non working days API Specification

cbliard opened the pull request:
https://community.openproject.org/work_packages/41816

Drafting the API
previous design:
```
- [x]    GET /api/v3/work_schedule/week_days
- [x]  PATCH /api/v3/work_schedule/week_days
- [x]    GET /api/v3/work_schedule/week_days/{day} # day 1-7
- [x]  PATCH /api/v3/work_schedule/week_days/{day}
- [x]    GET /api/v3/work_schedule/non_working_days # specific only
- [x]    GET /api/v3/work_schedule/non_working_days?include_weekend # recurring + specific
- [x]   POST /api/v3/work_schedule/non_working_days # create specific
- [x]    GET /api/v3/work_schedule/non_working_days/{date} # date ~ 2022-05-06
- [x]  PATCH /api/v3/work_schedule/non_working_days/{date} # date ~ 2022-05-06
- [x] DELETE /api/v3/work_schedule/non_working_days/{date} # date ~ 2022-05-06

# another iteration:
# Fetch non working days (both weekdays and specific). A filter can be added for the two types (e.g. {"type": { "operator": "=", "values": "NonWorkingDay"}}) or for the interval 
- [x] GET              /api/v3/days/work_schedule

```

current design:
```
# Manage and fetch the days of the week
- [x] GET/PATCH        /api/v3/days/week
- [x] GET/PATCH        /api/v3/days/week/{day} # day 1-7

# Manage and fetch the specific non working days
- [x] POST             /api/v3/days/non_working
- [x] GET/PATCH/DELETE /api/v3/days/non_working/{date} # date ~ 2022-05-06

# Fetch days (both working and non-working) with a link to the nonWorkingReasons. A filter can be added for the two types (e.g. {"working": { "operator": "=", "values": "true"}}) or for the interval 
- [x] GET              /api/v3/days
- [x] GET              /api/v3/days/{date}
```

cbliard committed:
[#41816] Non working days API

https://community.openproject.org/work_packages/41816

cbliard committed:
reuse existing code to get openapi spec yaml in validate_spec script

cbliard committed:
Add redocly configuration file

cbliard committed:
add patch /work_package/week_days

ulferts commented:
I only took a look at the structure, not at the specifics. The general distinction between week days and specific non working days makes sense to me and combining them into one end point to query them together sounds good.

Some of the routes I'd do differently just to be more flexible from a semantic standpoint. We might need the days for different purposes later on. Because of that I suggest to do:

```

# Manage and fetch the days of the week
GET/PATCH /api/v3/days/week
GET/PATCH /api/v3/days/week/{id} # id 1-7

# Fetch non working days (both weekdays and specific). A filter can be added for the two types (e.g. {"type": { "operator": "=", "values": "NonWorkingDay"}}) or for the interval 
GET /api/v3/days/work_schedule

# Manage and fetch the specific non working days
GET/PATCH/DELETE /api/v3/days/non_working/{date} # date ~ 2022-05-06
GET /api/v3/days/non_working
```

When more than one weekday is updated, we should use the self href as a reference:

```
{
  "type": "Collection":
  "_embedded": 
     "elements": [{
       "working": "false"
       "_links": {
         "self": {
           "href": "/api/v3/days/week/5"
         }
       }
     },
     ...
     ]
  }
  
  ```

Kharonus commented on .redocly.yaml:
```diff
@@ -0,0 +1,9 @@
+  extends:
+    - recommended
+  rules:
+    no-ambiguous-paths: off
```
Should we really turn this off? ambiguous paths are usually something, that hints to bad API design. And yes, we currently have some ambiguous paths. I actually hoped, that we eventually can get rid of those.

did you check, if this config file goes well with the `.redocly.lint-ignore.yaml`

cbliard replied:
I tried the vscode plugin for redocly and it gave a better experience than RubyMine, so I sticked with it, and it needed a `.redocly.yaml` file. When absent, redocly outputs this:

```
No configurations were defined in extends -- using built in recommended configuration by default.
Warning! This default behavior is going to be deprecated soon.
```

So I think it is compatible with it. Maybe I should have put my exceptions in that file.

I agree with ambiguous paths being bad design. But wanted to get rid of the inaccurate warnings in the openapi yaml file:

```yaml
  "/api/v3/queries/columns/{id}":
    "$ref": "./paths/queries_column.yml"
  "/api/v3/queries/{id}/star": # Paths should resolve unambiguously. Found two ambiguous paths: `/api/v3/queries/columns/{id}` and `/api/v3/queries/{id}/star`
    "$ref": "./paths/query_star.yml"
```

It is not smart enough to detect that `id` can only be integer and thus it will never be ambiguous.

I'd love to find a programmatic way to make it smarter about those ambiguous paths.

Kharonus replied:
Well, the path is still ambiguous in your example, even if `id` is an integer. Problem here is, that this can only be evaluated AFTER path resolvement (there are a lot of frameworks, that actually do not work like rails with grape here and already fail, when trying to build a backend with those pathes).

cbliard replied:
ok, I'll remove the rule exclusion then.

Kharonus commented on .editorconfig:
```diff
@@ -325,6 +325,7 @@ ij_xml_line_comment_at_first_column = true
 
 [{*.yml.example,*.yml,*.yaml}]
 indent_size = 2
+insert_final_newline = true
```
Could be a `[*]` rule, couldn't it? which extension is not supposed to end with a final newline?

cbliard replied:
Totally agree, and that's why I was asking for people against changing the rule in `[*]`. It may have been put there for a reason. I'll probably change it at some point.

Kharonus commented on script/api/validate_spec:
```diff
@@ -2,26 +2,21 @@
-spec_path = Pathname(__dir__).join("../../docs/api/apiv3/openapi-spec.yml")
-full_spec = API::OpenAPI.send(:assemble_spec, spec_path).to_yaml
+openapi_yaml_spec_path = "./tmp/oas-generated.yml"
+full_spec = `#{Pathname(__dir__).join('spec')} --format yaml`
```
I don't get it, why does this work? before the entry point of the OpenAPI spec was defined, now it is not. How does this script know, which spec file to validate?

cbliard replied:
Its's a refactoring. I just removed some duplication.

`spec` is a script that already exists and outputs the bundled openapi spec. At first I was commenting out the `FileUtils.rm` to get the bundled file, and then I saw the `spec` script, and thought that if the first script was calling the other, I would have discovered the existence of that `spec` script and I could use it without modifying the `validate_spec` script.

Since then I discovered that `npx @redocly/openapi-cli bundle docs/api/apiv3/openapi-spec.yml --output bundled.yaml` does exactly the same, without needing some code from us. Not sure if I should change it again.

Kharonus replied:
ahhhhhh, `spec` is the script, not the directory. If it works, I'm fine with that change.

Kharonus commented on docs/api/apiv3/paths/work_schedule_non_working_days.yml:
```diff
@@ -0,0 +1,106 @@
+          Example: `{ "interval": { "operator": "<>d", "values": ["2022-05-02","2022-05-26"] } }` would return non working days
+          between May 5 and May 26 2022, inclusive.
+      example: '[{ "interval": { "operator": "<>d", "values": ["2022-05-02","2022-05-26"] } }]'
+      required: false
```
`filter` is optional query param, but how does this endpoint behaves if the filter is not provided? It should go to the description, which currently only says, that non-working days for an interval are returned.

cbliard replied:
Yep good point. That's not specified for now.

cbliard replied:
It has been added. Thanks.

Kharonus commented on docs/api/apiv3/paths/work_schedule_non_working_days.yml:
```diff
@@ -0,0 +1,106 @@
+---
```
we usually put the path here as a comment, to make reviewing this file more easy.

cbliard replied:
Thanks. Done.

Kharonus commented on docs/api/apiv3/paths/work_schedule_non_working_days.yml:
```diff
@@ -0,0 +1,106 @@
+    Marks a day as being a non-working day.
+
+    Does nothing if the day is already a non-working day.
+  parameters:
```
is there a specific reason why to put the request data into mandatory query params, instead of using the request body? I'd say, the body should be the first choice when transporting creation request data.

cbliard replied:
That's because I learnt only very recently about `requestBody` :)
Thanks for pointing it out. I'll change it.

cbliard replied:
Done

Kharonus commented on docs/api/apiv3/components/schemas/week_day_write_model.yml:
```diff
@@ -0,0 +1,29 @@
+    type: integer
+    description: |-
+      The week day from 1 to 7. 1 is Monday. 7 is Sunday.
+    readOnly: true
```
this is a write model. it should not contain readonly properties.

Kharonus replied:
on second thought, the only writable property here is `working`. Hence, this whole write model should be nothing more then:

```
{
  working: true
}
```

Kharonus replied:
UPDATE: I saw, that you want to provide both, a single day PATCH and a week PATCH. For the later, of course you need `day` property in here, but still it is not a readonly property.

My question now would be: Why maintaining two endpoints? you could levy the week PATCH to work also for single days. (a collection with only one element.)

cbliard replied:
Not sure about it. I think the endpoints will make sense only once the UI will be developed. If the UI has a save button, then we need to save all at once, so PATCH all. If it updates each time you click on a check box, then it needs to update only one, and I thought it would be "cleaner" to patch the resource directly.

And I did not think about patching with a collection with only a single element too.

For the `readonly`, I probably left it after a copy-paste. Thanks for noticing.

And for the patch, could schema just be:
```
type: array
description: |-
  The array of week days.
minItems: 7
maxItems: 7
items:
  $ref: './week_day_write_model.yml'
```
instead of
```
type: object
required:
  - _type
  - _embedded
properties:
  _type:
    type: string
    enum:
      - Collection
  _embedded:
    type: object
    required:
      - elements
    properties:
      elements:
        type: array
        description: |-
          The array of week days.
        minItems: 7
        maxItems: 7
        items:
          $ref: './week_day_write_model.yml'
```

And as you suggest making it possible to patch only one day, I should set `minItems` to 1.

Kharonus replied:
I am indeed a little bit of a fan for batch endpoints :D Especially, because they fit both use cases OOTB, patching a single resource and patching multiple ones. ;) But honestly, I wouldn't mind supporting both, as they seem to have only minor overhead.

cbliard replied:
should be good with 2e2b103

Kharonus commented on docs/api/apiv3/paths/work_schedule_week_day.yml:
```diff
@@ -0,0 +1,126 @@
+        type: integer
+        minimum: 1
+        maximum: 7
+    - name: working
```
same here about query param vs. request body. In addition, you wrote a `WeekDayWriteModel`, but you do not use it here. (yet)

cbliard replied:
Nice spot

cbliard replied:
Done

Kharonus commented on docs/api/apiv3/paths/work_schedule_week_day.yml:
```diff
@@ -0,0 +1,126 @@
+            _type: Error
+            errorIdentifier: urn:openproject-org:api:v3:errors:InvalidQuery
+            message: You are not authorized to access this resource.
+    '422':
```
arguably we could skip 422 errors and return 404 here, if the index is not valid, because the index is a path param.

`working` property should not run into property constraint violations either, because anything beyond true and false is captured by a 400.

cbliard replied:
So I should remove the query parameter and replace it with a `requestBody` having the `WeekDayWriteModel` as schema.

cbliard replied:
good point for the 404

cbliard replied:
and the 400

cbliard replied:
Done in cda99dab51

Kharonus commented on docs/api/apiv3/paths/work_schedule_week_days.yml:
```diff
@@ -0,0 +1,99 @@
+      application/json:
+        schema:
+          $ref: ../components/schemas/week_day_collection_write_model.yml
+        example:
```
no example needed, if provided in the model file

cbliard replied:
Ok. I was not sure about how it would render on the website. Is there a waty to generate the same output as on https://www.openproject.org/docs/api/endpoints ?

Kharonus replied:
try http://localhost:4200/api/docs. it is not the same, but same enough ;) There is no polishing in this docs, like we have it at the website in the end.

cbliard replied:
Yep, it is same enough I guess. I'll probably ask Andrej if I need a more accurate one.

cbliard replied:
yep, done.

Kharonus reviewed (commented):
I have a couple of comments, some are questions, some improvement suggestions.

I only have one mayor discussion item about the PATCH for week days.

cbliard commented:
> I only took a look at the structure, not at the specifics. The general distinction between week days and specific non working days makes sense to me and combining them into one end point to query them together sounds good.
> 
> Some of the routes I'd do differently just to be more flexible from a semantic standpoint. We might need the days for different purposes later on. Because of that I suggest to do:
> 
> ```
> 
> # Manage and fetch the days of the week
> GET/PATCH /api/v3/days/week
> GET/PATCH /api/v3/days/week/{id} # id 1-7
> 
> # Fetch non working days (both weekdays and specific). A filter can be added for the two types (e.g. {"type": { "operator": "=", "values": "NonWorkingDay"}}) or for the interval 
> GET /api/v3/days/work_schedule
> 
> # Manage and fetch the specific non working days
> GET/PATCH/DELETE /api/v3/days/non_working/{date} # date ~ 2022-05-06
> GET /api/v3/days/non_working
> ```

Yep, makes sense. I'll change it to match your proposition

> When more than one weekday is updated, we should use the self href as a reference:
> 
> ```
> {
>   "type": "Collection":
>   "_embedded": 
>      "elements": [{
>        "working": "false"
>        "_links": {
>          "self": {
>            "href": "/api/v3/days/week/5"
>          }
>        }
>      },
>      ...
>      ]
>   }
>   
> ```

Ok, I'll use the href reference.

cbliard committed:
add GET/PATCH/DELETE /api/v3/non_working_days/{date}

cbliard commented:
@ulferts But this one seems a bit redundant though:

```
GET /api/v3/days/non_working
```

It can be done with proper filters using
```
GET /api/v3/days/work_schedule
```

Why should we keep both?

Edit: I know, you meant `POST`, right?

cbliard committed:
keep no-ambiguous-paths rule active

cbliard committed:
make api scripts work from any working directory

cbliard committed:
use more flexible and semantic routes scheme

We may also need the days for different purposes later on.

cbliard committed:
dont validate openapi spec if bundling failed

cbliard committed:
put parameters inside the request body

cbliard committed:
identify a week day by its href

also remove the day from the WeekDay write model as it is already present in the query

cbliard committed:
use 404 and 400 instead of 422 for PATCH /api/v3/days/week/{day}

cbliard commented:
@oliverguenther @ulferts @Kharonus I would need some comments about an endpoint.

Semantically, I feel like having `GET /api/v3/days/work_schedule` returning non working days is wrong. I would expect it to return the work schdule, so both working and non-working days.

Maybe it could return an object like this:
```
_type: WorkSchedule
from: '2022-05-01'
to: '2022-06-30'
workingDays: [
  '2022-05-02',
  '2022-05-03',
  '2022-05-04',
  ...
],
nonWorkingDays: [
  '2022-05-01',
  '2022-05-07',
  '2022-05-08',
  ...
]
_links:
  detail:
    href: '/api/v3/days/work_schedule/{date}'
    templated: true
```
Endpoint `/api/v3/days/work_schedule/{date}` can be requested to get more information about a particular date. It would return all known information about that particular date.

But then it would be cumbersome to get the day names. So maybe it could be this instead:
```
_type: WorkSchedule
from: '2022-05-01'
to: '2022-06-30'
workingDays: [
  {
    _type: WeekDay,
    date: '2022-05-02',
    name: Monday
  }, {
    _type: WeekDay,
    date: '2022-05-03',
    name: Tuesday
  }, {
    _type: WeekDay,
    date: '2022-05-04',
    name: Wednesday
  },
  ...
],
nonWorkingDays: [
  {
    _type: WeekDay,
    date: '2022-05-01',
    name: Sunday
  }, {
    _type: NonWorkingDay,
    date: '2022-05-01',
    name: Labour day
  }, {
    _type: WeekDay,
    date: '2022-05-07',
    name: Saturday
  }, {
    _type: WeekDay,
    date: '2022-05-08',
    name: Sunday
  },
  ...
]
_links:
  detail:
    href: '/api/v3/days/work_schedule/{date}'
    templated: true
```
Notice `2022-05-01` appearing twice

But it would be wrong to use `WeekDay` and `NonWorkingDay` here, because I want to have only `date` and `name` here, to keep it succinct, and the current schema is defining some other mandatory fields. So maybe it could be a `Day` model with an optional `nonWorkingDayName` like this instead:

```
_type: WorkSchedule
from: '2022-05-01'
to: '2022-06-30'
workingDays: [
  {
    _type: Day,
    date: '2022-05-02',
    weekDayName: Monday
  }, {
    _type: Day,
    date: '2022-05-03',
    weekDayName: Tuesday
  }, {
    _type: Day,
    date: '2022-05-04',
    weekDayName: Wednesday
  },
  ...
],
nonWorkingDays: [
  {
    _type: Day,
    date: '2022-05-01',
    weekDayName: Sunday,
    nonWorkingDayName: Labour day
  }, {
    _type: Day,
    date: '2022-05-07',
    weekDayName: Saturday
  }, {
    _type: Day,
    date: '2022-05-08',
    weekDayName: Sunday
  },
  ...
]
_links:
  detail:
    href: '/api/v3/days/work_schedule/{date}'
    templated: true
```

I have the feeling that it would be easier to consume from a client perspective, but I may be wrong.

What do you think? Is it worth it? Any other ideas to make it more correct?

Kharonus commented:
I'd rather not bake a custom bread on this endpoint. Instead it should return a simple HAL collection of some `Day` objects. Those embedded elements then contain all the information a client need. The client can filter, sort, or group, whatever he wants. I'm thinking now, you should embrace that idea to create a single model `Day` which embodies all information. A `Day` resource is linked to a week day and optional to a non-working day, hence you have the information of those in `_links`. The `working` property can be derived from any `NonWorkingDay` with that `date` and the `WeekDay`, which can be marked as `working: false`. (combining them with an `&` operator, so if any is `false` then the `Day` object is `working: false`)

Like this:

```
example:
  _type: Collection
  total: 6
  count: 6
  _links:
    self:
      href: '/api/v3/days/work_schedule/?filters=[{"interval":{"operator":"<>d","values":["2022-05-01","2022-05-05"]}}]'
  _embedded:
    elements:
      - _type: Day
        date: '2022-05-01'
        working: false
        _links:
          self:
            href: /api/v3/days/work_schedule/2022-05-01
            title: Labour day
          weekDay:
            href: /api/v3/days/week/7
            title: Sunday
      - _type: Day
        date: '2022-05-02'
        working: true
        _links:
          self:
            href: /api/v3/days/work_schedule/2022-05-02
          weekDay:
            href: /api/v3/days/week/1
            title: Monday
      - ...
```

I'm not 100% satisfied yet, because of the various resource types we have here:

- WeekDay: a generic day resource for every day of the week (1-7)
- NonWorkingDay: a singular resource defined by the date, that marks the very day as `working: false`
- Day: a representational resource of any day in the year (work schedule), which combines information from both above.

cbliard commented:
Thanks @Kharonus. I tend to forget we must return valid HAL. Your example work, and I think the `Day` resource can also have some display information.

If someone consumes the API to display a calendar, they would not know which name to display for non-working days. I feel like it would be wrong to look up the name in the `_links` section, but I may be wrong. And displaying information for the day of the week is so obvious that I do not think we need to embed it in the model (contrary to what I proposed).

I agree with the feeling of not being 100% satisfied.

cbliard committed:
fix GET /api/v3/days/work_schedule and hrefs

Kharonus commented:
Generally I see no problem with reading information from the links. That's what they are for. ;)

cbliard commented:
I linked to the wrong work package. There is #41816 which is API implementation. I'll link to this one instead.

dombesz commented on docs/api/apiv3/components/schemas/non_working_day_model.yml:
```diff
@@ -0,0 +1,40 @@
+              This non working day. If the non-working day is recurring like a weekend day, the resource will point to a
+              week day.
+
+              **Resource**: NonWorkingDay or WeekDay
```
Shouldn't the resource be a `NonWorkingDay` only?

cbliard replied:
Well, we had this distinction between a recurring non-working day, which is a `WeekDay` with `working == false`, like Sundays, and the non-working days being set manually and ever changing, like the Easter Monday and it is represented by a `NonWorkingDay`. At some point, @Kharonus brought the idea of having only a `Day`, with a `working` attribute, and it would link to either the `NonWorkingDay` or the `WeekDay`. Maybe a `reason` link, or something similar.

But you are right. If it is a `NonWorkingDay` model, and it links to self, then it should link to the same `NonWorkingDay`

cbliard replied:
With 314190da32, resource is now a `Day`, and it has `working` set to `false`, it will link to `NonWorkingDay` and/or `WeekDay` describing why in its `nonWorkingReasons` link.

dombesz commented on docs/api/apiv3/paths/days_non_working.yml:
```diff
@@ -0,0 +1,48 @@
+# /api/v3/days/non_working
+---
+post:
+  summary: Create a non working day
```
Is it possible to create a recurring non-working day? I'm thinking of recurring public holidays, 1st of May, etc.

cbliard replied:
This is out-of-scope for now. Not sure if that will be planned or not.

dombesz commented on docs/api/apiv3/components/schemas/non_working_day_model.yml:
```diff
@@ -0,0 +1,40 @@
+        allOf:
+          - $ref: './link.yml'
+          - description: |-
+              This non working day. If the non-working day is recurring like a weekend day, the resource will point to a
+              week day.
```
```suggestion
              This is a non working day. If the non-working day is recurring like a weekend day, the resource will point to a
              week day.
```

I'm not sure I understand the second sentence, what is the difference between a non-working day and a weekday?

cbliard replied:
It should be read "Links to this non-working day" rather than "This non-working day". It is the description of the link. And it refers to the fact that there are multiple reasons for a dayu to be a non-working one.

How would you rephrase it?

cbliard replied:
It has changed with 314190da32.

dombesz commented on docs/api/apiv3/paths/days_non_working_date.yml:
```diff
@@ -0,0 +1,133 @@
+  description: |-
+    Returns the non-working day information for a given date.
+  parameters:
+    - name: date
```
How do we know whether the non-working day is a recurring one?

cbliard replied:
For this endpoint, it is only about setting and retrieving the non-recurring non-working days.

dombesz commented on docs/api/apiv3/paths/days_non_working_date.yml:
```diff
@@ -0,0 +1,133 @@
+    - Work Schedule
+    - Non working days
+  description: |-
+    Removes the day of the given day from the non-working days known.
```
Not sure what does this mean.

cbliard replied:
If you did a typo and entered 15th December as a non-working day for Christmas, then you want to delete it and create a new one (or update it, but let's pretend that's not possible). To delete it this endpoint should be used.

The sentence reads weirdly I agree. How would you rephrase it?

cbliard replied:
Fixed in 54ee9fc31b

Is the new wording better?

cbliard replied:
It has changed with 314190da32

dombesz commented on docs/api/apiv3/paths/days_non_working_date.yml:
```diff
@@ -0,0 +1,133 @@
+            errorIdentifier: urn:openproject-org:api:v3:errors:NotFound
+            message: The requested resource could not be found.
+
+# TODO: add missing description, missing tags, missing parameters, and missing response codes
```
There is a todo here.

cbliard replied:
Thanks, I forgot to remove it.

dombesz commented on docs/api/apiv3/paths/days_non_working_date.yml:
```diff
@@ -0,0 +1,133 @@
+      content:
+        application/hal+json:
+          schema:
+            $ref: '../components/schemas/error_response.yml'
```
Not sure if we can do something about this, as not directly related to this specification. But, the example value for the 400 response seems to be off. This is just a nitpick.

<img width="1187" alt="image" src="https://user-images.githubusercontent.com/83396/168070271-23d1bdb3-65f5-47e6-8c27-060e9808db55.png">

cbliard replied:
Fixed in 6930c15c33.

dombesz commented on docs/api/apiv3/paths/days_week.yml:
```diff
@@ -0,0 +1,100 @@
+          schema:
+            $ref: '../components/schemas/week_day_collection_model.yml'
+    '400':
+      description: Returned if the client sends invalid request parameters
```
This request has no request parameters. Perhaps we should rather add a `403 Returned if the client does not have sufficient permissions.`

cbliard replied:
You're right, it's hard to get a 4xx error here.
Maybe 403 if we restrict these resources to logged-in users only.

cbliard replied:
This endpoint should be publicly available, so no 403.
If we restrict to logged in users, that means frontend will not be able to display elements of public projects properly (Gantt chart, calendars, etc.) because the week days and non-working days information can not be fetched.

dombesz commented on docs/api/apiv3/paths/days_non_working_date.yml:
```diff
@@ -0,0 +1,133 @@
+        application/hal+json:
+          schema:
+            $ref: '../components/schemas/error_response.yml'
+    '404':
```
Should we also add a `403 The client does not have the needed permissions to perform the requested action` in case the user is not logged in ? Would that be useful for all the endpoints that require login?

cbliard replied:
Maybe. That's the kind of things I was not sure about.

@ulferts do you think some of the days endpoints can be public, or they at least need to be logged-in?

cbliard replied:
This endpoint should be publicly available, so no 403.
If we restrict to logged in users, that means frontend will not be able to display elements of public projects properly (Gantt chart, calendars, etc.) because the week days and non-working days information can not be fetched.

dombesz commented on docs/api/apiv3/paths/days_week_day.yml:
```diff
@@ -0,0 +1,125 @@
+  description: |-
+    Makes a week day a working or non-working day.
+
+    TODO: describe what should happen if a working attribute of a week day is changed and the impact it will have on
```
There is a todo here.

cbliard replied:
the behavior is now specified in [6a9dd71](https://github.com/opf/openproject/pull/10650/commits/6a9dd7191a756a97e69373355feede6a8bbda598).

dombesz commented on docs/api/apiv3/paths/days_week_day.yml:
```diff
@@ -0,0 +1,125 @@
+            _type: Error
+            errorIdentifier: urn:openproject-org:api:v3:errors:InvalidRequestBody
+            message: The request body was not a single JSON object.
+    '422':
```
There is a bit of a status inconsistency between this '422' and the [`PATCH /api/v3/days/week/{day}`](https://github.com/opf/openproject/pull/10650/files#diff-1d1e4b5f246f5f58acd252ac654e4cd577dad49bead730608e8e62bd5cd78469R54)
This request will return a '422' when the day parameter is out of the range 1-7, and the PATCH will return a 404 error.

I think it would make sense for all the requests (GET, POST) to return a 404, when the day is out of range.
The other requests are also returning '404' when the range is out of 7.

dombesz replied:
I would also keep the 422, for the cases when there is a validation error.

Furthermore, the standard rails way is to return a 422 whenever the create, update actions have a validation error.
I would propose a 422 for the following actions too:
- POST /api/v3/days/non_working Create a non working day
- PATCH /api/v3/days/non_working/{date} Update a non-working day attributes
- PATCH /api/v3/days/week Update week days
- PATCH /api/v3/days/week/{day} Update a week day attributes

cbliard replied:
I changed from 422 to 404 on @Kharonus suggestion, but I did not update the other similar endpoints. That's a mistake from me. You're right, this should be consistent and return 404 everywhere in this case.

For the 422, if there is a validation error, it should be covered by a 400: the request is invalid. I think that's how the rest of the API currently behaves. Do you have some reference about the standard rails way to return a 422 whenever the create, update actions have a validation error?

cbliard replied:
Fixed in 54ee9fc31b, and removed some 422. Still not sure which one should be used between 400 and 422.

dombesz replied:
Sorry, I cannot find a reference to 422 being the standard rails way. However, [rails g scaffold command](https://github.com/rails/rails/blob/a8f3f0df0ad26c3b6f614af896538982dcf329f0/railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb.tt#L24 ) will generate a controller that uses 422 :unprocessable_entity for handling validation errors. I'm also fine with 400 if that's what we are using through the api.

dombesz commented on docs/api/apiv3/paths/days_non_working.yml:
```diff
@@ -0,0 +1,48 @@
+# /api/v3/days/non_working
```
There are a lot of duplicate routes between the different titles:
<img width="1203" alt="image" src="https://user-images.githubusercontent.com/83396/168081945-40bec3ca-d3d8-4aef-bb19-3aa2fff60f58.png">

<img width="1233" alt="image" src="https://user-images.githubusercontent.com/83396/168082138-07512029-7c18-4555-9c32-eec2b8ba04de.png">
<img width="1235" alt="image" src="https://user-images.githubusercontent.com/83396/168082183-cfa0d9b6-4c56-49d0-9ba3-d01bd5d113cc.png">
<img width="1239" alt="image" src="https://user-images.githubusercontent.com/83396/168082219-fc4c44d9-bd1d-4e5a-a92c-fa64418fd431.png">

cbliard replied:
I think that's me not knowing how the OpenAPI `tags` should be used. I'll have to fix it. And also add a description to these tags.

cbliard replied:
Fixed in 54ee9fc31b

There is only one "Work Schedule" section now with a brief description of the intended behavior.

dombesz commented on docs/api/apiv3/paths/days_week_day.yml:
```diff
@@ -0,0 +1,125 @@
+                message: The request was invalid. Expected day to be between 1 and 7.
+
+patch:
+  summary: Update a week day attributes
```
Whenever we submit the body `{ "_type": "WeekDay", "working": false }`, does the `"_type"` becomes a `NonWorkingDay`?

cbliard replied:
No, it is used to mark week days as being worked on or not. Then when a work schedule is requested for a period of time, it will return non-working days. Some of them will be non-working due to being non-working week days (weekend days) and some other because manually added non-working days.

dombesz commented on docs/api/apiv3/paths/days_non_working_date.yml:
```diff
@@ -0,0 +1,133 @@
+  description: |-
+    Removes the day of the given day from the non-working days known.
+
+    Trying to remove a date which is not a non-working day, by calling the same
```
What is the reason behind this behaviour?

cbliard replied:
I thought it would be better but [I'm probably wrong](https://stackoverflow.com/questions/6439416/status-code-when-deleting-a-resource-using-http-delete-for-the-second-time), and returning 404 would probably be better.

cbliard replied:
Fixed in 54ee9fc31b

dombesz reviewed (changes requested):
Some suggestions and a few questions.

cbliard committed:
remove todo comment

[skip ci]

cbliard committed:
update example of invalid query

cbliard committed:
Add work schedule description + use 404 where appropriate

ulferts commented:
> @ulferts But this one seems a bit redundant though:
> 
> ```
> GET /api/v3/days/non_working
> ```
> 
> It can be done with proper filters using
> 
> ```
> GET /api/v3/days/work_schedule
> ```
> 
> Why should we keep both?
> 
> Edit: I know, you meant `POST`, right?

Yes, that's what I meant. Sorry for the confusion.

ulferts commented:
@cbliard , @Kharonus this is my idea for how a collection of days could look like:

```

{
  "type": "Collection",
  "links": {
    "self": {
      "href": '/api/v3/days?filters=[{"interval":{"operator":"<>d","values":["2022-05-01","2022-05-02"]}}]'
    }
  },
  "count": 2,
  "total": 2,
  "_embedded": {
    "elements": [
      # The Day objects are non writable
      {
        "_type": "Day",
        # Would not actually be necessary since whenever there is a nonWorkingReason, it will
        # be a non working day. But it is explicit.
        "working": "false",
        # The title is constructed for the client based on the weekday and the non working day
        # if that exists. If the client disagrees with that way of displaying the information,
        # the raw information can be fetched form the "nonWorkingReasons"
        "title": "Sunday (Labor day)",
        "date": "2022-05-01",
        "_links": {
          "self": {
            "href": "/api/v3/days/2022-05-01"
          },
          "nonWorkingReasons": [
            {
              href: "api/v3/days/week/7",
              title: "Sunday"
            },
            {
              href: "api/v3/days/non_working/2022-05-01",
              title: "Labor Day"
            }
          ],
          "weekday": {
            href: "api/v3/days/week/7",
            title: "Sunday"
          }
        }
      },
      {
        "_type": "Day",
        # Would not actually be necessary since whenever there is no nonWorkingReason, it will
        # be a working day. But it is explicit.
        "working": true,
        "title": "Monday",
        "date": "2022-05-02",
        "_links": {
          "self": {
            "href": "/api/v3/days/2022-05-02"
          },
          "nonWorkingReasons": [],
          "weekday": {
            href: "api/v3/days/week/1",
            title: "Monday"
          }
        }
      }
    ]
  }
}
```

So it is very similar to @Kharonus proposal but with a few changes. For one it sports the `nonWorkingReasons` with potential references to both the weekday as well as to a specific non working day. That way, the client can reason about why the day is a non working day and the user can perform the necessary changes to alter the state if deemed necessary. Otherwise, the user might be left wondering. 

It also alters the routes to not include `work_schedule`. If I understand [the specification on path objects](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#pathsObject) correctly, `/api/v3/days`, `/api/v3/days/{date}`, `/api/v3/days/week/{id}`, `/api/v3/days/week`, `/api/v3/days/non_working` and `/api/v3/days/non_working/{date}` would not be considered ambiguous because of the way concrete definitions are preferred over templated. This also broadens the semantic of what the `Day` can be used for.

I for one don't think having three types (`Day`, `Weekday` and `NonWorkingDay`) is bad. They each serve a distinct purpose. Trying to merge them into one type would fail I think. E.g. `NonWorkingDay` might get a `repeating` property. And the `Weekday` is not a concrete day and as such does not have a `date` property. 

One thing that we might want to consider is to allow filtering the `api/v3/days` endpoint by the `working` attribute since this is what at least our client will be interested in most.

cbliard commented:
@ulferts I think it adds up on our previous discussions with @Kharonus and merge the concerns we had.

I'll change the `days/work_schedule` into `days` and `days/{date}` as you both proposed. Thanks

cbliard committed:
specify /api/v3/days and /api/v3/days/{date} endpoints

cbliard committed:
add description and permissions to /api/v3/days/* endpoints

cbliard committed:
specify behavior when marking a day as non-working

as specified by Parimal in https://community.openproject.org/projects/openproject/work_packages/18416/activity\#activity-116 it will not move dates, but rather change duration of work packages.
